llmfoundry/data/finetuning/tasks.py

```diff
@@ -332,6 +333,13 @@ def build_from_hf(
             preprocessing_fn = self.get_preprocessing_fn_from_str(
                 proto_preprocessing_fn, dataset_name)
 
+        signal_file_path = f'.node_{dist.get_node_rank()}_local_rank0_data_prep_completed'
+
+        if dist.get_local_rank() != 0:
```

Comment:
Question: Do you need this line since `local_rank_zero_download_and_wait()` is already waiting for non-local ranks?

Reply:
yes, the way this works is in multiple steps. First, all non local rank 0s wait at `local_rank_zero_download_and_wait()` for the `signal_file` to be written. local rank 0s proceed with all of the dataset processing. This is a file barrier that will not time out. Then, once local rank 0 has finished processing the dataset, it writes the signal file. Then all ranks proceed to the `barrier`. Then, local rank0s exit this code while non local rank 0s load the already processed datasets (should be pretty much instant). Finally, non local rank0s exit this code.